File: clients/kotlin/ridesharesdk/src/main/kotlin/rideshare/sdk/Rider.kt
package rideshare.sdk

import ares.client.fhe.CryptoContext

object Rider {
    fun keygen(
        ringDim: Int = 1 shl 15, depth: Int = 5,
        scalingFactor: Double = Math.scalb(1.0, 50)
    ): Keypair {
        val ctx = CryptoContext(ringDim, scalingFactor, depth)
        ctx.use {
            val kp = ctx.singleKeyGen()
            return Keypair(ctx.serialize(kp.publicKey), ctx.serialize(kp.secretKey))
        }
    }

    fun decryptMasks(
        skBytes: ByteArray, encryptedMasks: List<ByteArray>,
        ringDim: Int = 1 shl 15, depth: Int = 5,
        scalingFactor: Double = Math.scalb(1.0, 50)
    ): Pair<Int, DoubleArray> {
        val ctx = CryptoContext(ringDim, scalingFactor, depth)
        ctx.use {
            val sk = ctx.deserializeSecretKeyShare(skBytes, true)
            val masks = DoubleArray(encryptedMasks.size)
            var best = -1; var bestVal = 0.0
            for ((i, ctBytes) in encryptedMasks.withIndex()) {
                val ct = ctx.deserializeCiphertext(ctBytes)
                val vals = ctx.decryptSingle(ct, sk, 1)
                masks[i] = vals[0]
                if (best < 0 || vals[0] > bestVal) { best = i; bestVal = vals[0] }
            }
            return Pair(best, masks)
        }
    }
}

File: clients/rideshareswift/Sources/RideshareCoreLib/Rider.swift
import AresClientFHE

public enum Rider {
    public static func keygen(ringDim: UInt32 = 1 << 15, depth: UInt32 = 5,
        scalingFactor: Double = Double(UInt64(1) << 50)) throws -> Keypair {
        let ctx = try CryptoContext(ringDim: ringDim, scalingFactor: scalingFactor, depth: depth)
        let kp = try ctx.singleKeyGen()
        let pk = try ctx.serialize(kp.publicKey)
        let sk = try ctx.serialize(kp.secretKey)
        return Keypair(publicKey: Data(pk), secretKey: Data(sk))
    }

    public static func decryptMasks(skBytes: Data, encryptedMasks: [Data],
        ringDim: UInt32 = 1 << 15, depth: UInt32 = 5,
        scalingFactor: Double = Double(UInt64(1) << 50)) throws -> (winnerIndex: Int, maskValues: [Double]) {
        let ctx = try CryptoContext(ringDim: ringDim, scalingFactor: scalingFactor, depth: depth)
        let sk = try ctx.deserializeSecretKeyShare(Array(skBytes), lead: true)
        var masks = [Double](); var best = -1; var bestVal = 0.0
        for (i, ctBytes) in encryptedMasks.enumerated() {
            let ct = try ctx.deserializeCiphertext(Array(ctBytes))
            let vals = try ctx.decryptSingle(ct, with: sk, slots: 1)
            masks.append(vals[0])
            if best < 0 || vals[0] > bestVal { best = i; bestVal = vals[0] }
        }
        return (best, masks)
    }
}
